### DISCUSSION TITLE
DUAL LAN Problems

### DISCUSSION THREAD

**Discussion by storchalberto** [arendst/Tasmota#21158] [Support request] [answered]
I have a ESP32-ETH01 board. I heve read thet it is necessary to assign two different IP address for ethernet and wifi.
My configuration:
Ipaddress1 192.168.1.25
Ipaddress2 192.168.1.1
Ipaddress3 255.255.255.0
Ipaddress4 8.8.8.8
Ipaddress5 8.8.4.4
EthIPAddress 192.168.1.26
EthGateway 192.168.1.1 
EthSubnetmask 255.255.255.0 
EthDNSServer1 8.8.8.8
EthDNSServer2 8.8.4.4
EthAddress 1
Ethclockmode 0
Ethtype 0
The ESP32 respond on both the address. I am using the functionality to add my pages to the nweb server and i have tested it on a ESP32 standard board. Now i heve loaded the same berry program in the ESP32-ETH01 model.
The problem is that the device start always indicating the WEB server on the ip address of the Ethernet line.
So if i disconnect the ethernet cable, teh device respond with standard pages on the wifi address, but do not respond to my pages on it, indicating: "file not found".
I have used the berry script to do flip-flop for the connection, but it do not solve the problem.
If i stop the ethernet (ethernet 0) and do a restart with ethernet disabled, the web server start on the wifi address and all function well.
There is a way to solve this problem, i am using the device normally connected by ethernet, but using the wifi as backup connection.

**Comment by sfromis** [comment]
With WT32-ETH01 (which I assume that you mean), I'm seeing the Tasmota web server respond on both Wifi and Ethernet. I did not do any IP address configuration (except for having my DHCP set fixed IP addresses by MAC). Tested with both the standard Tasmota web UI server, and a Berry web server.

**Reply by s-hadinger** [reply]
Can you share your Berry code?

If you listen to "" (empty string), it listens automatically to all IP addresses.

**Reply by sfromis** [reply]
With a web server, there is no concept of "listening" to IP addresses, any or all. And I just don't care about what IP addresses is mentioned in the `HTP: Web server active` message, this is not an indication of where it was "listening", just one IP address available at that time. Not sure where you are even using an empty string....

Anyway, here's my trivially basic webserver test code:
```
class berryfs

  def init()
    tasmota.remove_driver(global.berryfs_instance)
    tasmota.add_driver(global.berryfs_instance := self)
  end

  def web_add_handler()
    import webserver
    webserver.on("/fs", / -> self.webrequest(), webserver.HTTP_GET)
  end

  def webrequest()
    import webserver
    try
      var fh = open(webserver.arg('file'),'r')
      webserver.content_response(fh.read())
      fh.close()
    except .. as e, m
      webserver.content_response('error:'..e..' '..m)
    end
  end

end

berryfs()
```

**Reply by barbudor** [reply]
> the device start always indicating the WEB server on the ip address of the Ethernet line

Where do you see that ?
IP address should NOT be be in any links to the device itself
All links in the UI should be relative

**Reply by storchalberto** [reply]
In attach
[TasmotaProblem.txt](https://github.com/arendst/Tasmota/files/14961185/TasmotaProblem.txt)

**Reply by sfromis** [reply]
In general, reading the log (at least level 3) carefully for details of what happens when is a key step to do debugging. Maybe your creation of the web server instance happened before the network connection was established. At least, your symptoms are consistent with no web server instance registered.

**Reply by storchalberto** [reply]
to do the testing i have inserted the code in a new berry script (testweb .be) and midified the autoexec.be in:
import webserver
def Avvio()
	print ("Carico TESTWEB")
	load ("testweb.be")
end
tasmota.set_timer(10000, Avvio)
print("Attesa Caricamento moduli")
so the creation is subsequent to the connection, as i see in the log:
00:00:04.512 Attesa Caricamento moduli
00:00:04.513 BRY: Successfully loaded 'autoexec.be'
00:00:06.001 WIF: Connecting to AP1 ******************* Channel 12 BSSId 80:16:05:F1:71:68 in mode HT20 as tasmota-B40778-1912...
00:00:08.253 WIF: Connected
00:00:08.486 HTP: Web server active on tasmota-B40778-1912 with IP address 192.168.1.25
00:00:11.954 QPC: Reset
00:00:13.045 RSL: INFO1 = {"Info1":{"Module":"WT32_ETH01","Version":"13.4.0.4(ESP32_ALLARME_ZB)","FallbackTopic":"cmnd/DVES_B40778_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:13.063 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-B40778-1912","IPAddress":"192.168.1.26","IP6Global":"","IP6Local":"fe80::c6de:e2ff:feb4:778%st2","Ethernet":{"Hostname":"tasmota-B40778-1912-eth","IPAddress":"192.168.1.25"}}}
00:00:13.091 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":46}}
00:00:15.072 Carico TESTWEB
00:00:16.615 RSL: STATE = {"Time":"1970-01-01T00:00:16","Uptime":"0T00:00:13","UptimeSec":13,"Heap":126,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":242,"MqttCount":0,"Berry":{"HeapUsed":5,"Objects":78},"Wifi":{"AP":1,"SSId":"VodafoneReteCasaFibra","BSSId":"80:16:05:F1:71:68","Channel":12,"Mode":"HT20","RSSI":46,"Signal":-77,"LinkCount":1,"Downtime":"0T00:00:04"}}
00:00:16.717 RSL: SENSOR = {"Time":"1970-01-01T00:00:16","ANALOG":{"A1":244,"A2":301}}
but continue to do not function until i send the command brrestart

**Reply by sfromis** [reply]
If you also had followed my suggestion of *reading* the log carefully, it should be clear where it did not do what you expected. The message "page added" which you fortunately did add to `web_add_handler` was not printed, hence this function was not executed. As the Berry/Tasmota docs explains, it will be "called when Tasmota web server started", which does not work with the delayed start, long after the web server was started.

Anyway, as the code example worked for me when not having this postponement of loading the driver, this means that it is not explained what caused the earlier failure for you. But **do** read the log carefully to check what happens, or not.

Your log also shows another bad thing which you did not mention at all, that you did not allow it to have real time (typically NTP server) access. Without this, Tasmota is not "fully initialized", and some functions may not work as expected. Whether this is part of your problems, I don't know.

**Reply by storchalberto** [reply]
In the attach the result of the test (restart 1 and then brrestart) on the seriallog (seriallog 4) , the webserver activation is directly in the autoexec.be file.
[TasmotaProblem1.txt](https://github.com/arendst/Tasmota/files/14965654/TasmotaProblem1.txt)

**Reply by storchalberto** [reply]
I think to have identified the problem, i attach the logs. If there is the ethernet interface enabled, it seems that the webservices starts at the wrong time. if i introduce a delay all runs well. if i deactivate the ethernet connection, there are no problems.
You think it is possible to identify and solve the problem or is necessary to do a test in the berry coding to avoid the problem (the delay
[TasmotaProblemLAN.txt](https://github.com/arendst/Tasmota/files/14967163/TasmotaProblemLAN.txt)
 solution obviously is not the best)?

**Reply by storchalberto** [reply]
I have modified the progrem:
  def web_add_handler()
    print(tasmota.wifi())
    webserver.on("/test1.htm", / -> self.page_say_test1())
    print("page added")
  end
In the LOG, if a do a restart 1 i see 2 times the execution of web_add_handler() and the print in both the cases is:
{'mac': 'C4:DE:E2:B4:07:78', 'up': false}
and the http access to the test1 page is nok.
If i do the brrestart, i see only une time in nthe log:
{'up': true, 'quality': 48, 'mac': 'C4:DE:E2:B4:07:78', 'rssi': -76, 'ip': '192.168.1.26', 'ip6local': 'fe80::c6de:e2ff:feb4:778%st2'}
and the and the http access to the test1 page is ok.
So i think that the web_add_handler in case of ethernet+wifi  is generated not at the corrett time.
Now i try to modify the program to test in the web add handler if the wifi connection is enabled and up and then do the weìbserver.on.

**Reply by storchalberto** [reply]
Tested to run the webserver.on("/test1.htm", self.page_say_test1) after the test of the connection is ok, but do not function. I heve verified that a delay of 3000mSec after the boot is not sufficient, 4000mSec it is OK.

**Reply by s-hadinger** [reply]
I'm on it now.

**Comment by s-hadinger** [comment, marked as answer]
Fixed in https://github.com/arendst/Tasmota/pull/21272

Thanks for reporting and taking time to isolate the problem.
